mini-projects/01-device-driver/laptop-interface/src/game_manager.rs

```diff
@@ -0,0 +1,413 @@
+                let mut input = String::new();
+                match self.stdin_receiver.try_recv() {
+                    Ok(user) => input = user.trim().to_string(),
+                    // If we don't have any user input, move on
+                    Err(_) => (),
+                }
+
+                // Determine action based on user input
+                match input.as_str() {
+                    "ready" => self.move_to_game(),
+                    // If there is no input, do nothing
+                    "" => (),
+                    _ => {
+                        warn!("Invalid Input: {}. Expected \"ready\".", input);
+                        println!("Invalid Input. Please enter \"ready\" to start the game.");
+                    }
+                }
+            }
+
+            GameState::Ingame => {
+                thread::sleep(std::time::Duration::from_millis(17));
+                // Actual game logic
+
+                // Check for new controller
+                match self.controller_manager.connect_new_controller() {
+                    Ok(Some(id)) => {
+                        let player_number = self.add_player(id);
+                        println!("Connected Player {}", player_number);
+                    }
+                    Ok(None) => (),
+                    Err(e) => {
+                        error!("Error Connecting Controller: {}", e);
+                        println!("Error Connecting Controller: {}", e);
+                    }
+                }
+                self.controller_manager.update_controller_state();
+                // Poll each player's controller
+                for player in self.players.iter_mut() {
+                    // Get the new state of their controller
+                    match self
+                        .controller_manager
+                        .get_controller_state(player.controller_id)
+                    {
+                        Some(state) => {
+                            // Convert controller state to movement
+                            let movement = controller_state_to_movement(state);
+                            player.position.0 += movement.0;
+                            player.position.1 += movement.1;
+
+                            // Update score based on position
+                            player.score = (player.position.0 as f64).powi(2)
+                                + (player.position.1 as f64).powi(2);
+
+                            // Add score to history if at least 5 seconds have passed
+                            if let Some(timer) = self.timer {
+                                if timer.elapsed().as_millis() >= 50 {
+                                    player
+                                        .history
+                                        .push((player.score, timer.elapsed().as_micros() as f64));
+                                    self.timer = Some(Instant::now());
```

Comment:
if you're trying to log the player's score over time, wouldn't constantly resetting the timer's starting time cause the elapsed time to not be monotonic?